codalab/model/bundle_model.py

```diff
@@ -539,113 +543,114 @@ def make_condition(key, field, value):
-                            and_(cl_bundle_metadata.c.metadata_key == key, condition)
-                        )
+                    add_join(
+                        aliased_bundle_metadata,
+                        cl_bundle.c.uuid == aliased_bundle_metadata.c.bundle_uuid,
                     )
 
-            if clause is not None:
-                clauses.append(clause)
+                    where_clause = and_(aliased_bundle_metadata.c.metadata_key == key, condition)
 
-        clause = and_(*clauses)
+            if where_clause is not None:
+                where_clauses.append(where_clause)
+
+        where_clause = and_(*where_clauses)
 
         if user_id != self.root_user_id:
             # Restrict to the bundles that we have access to.
+            add_join(
+                cl_group_bundle_permission,
+                cl_bundle.c.uuid == cl_group_bundle_permission.c.object_uuid,
+            )
             access_via_owner = cl_bundle.c.owner_id == user_id
-            access_via_group = cl_bundle.c.uuid.in_(
-                select([cl_group_bundle_permission.c.object_uuid]).where(
-                    and_(
-                        or_(  # Join constraint (group)
-                            cl_group_bundle_permission.c.group_uuid
-                            == self.public_group_uuid,  # Public group
-                            cl_group_bundle_permission.c.group_uuid.in_(
-                                alias(
-                                    select([cl_user_group.c.group_uuid]).where(
-                                        cl_user_group.c.user_id == user_id
-                                    )
-                                )
-                            ),  # Private group
-                        ),
-                        cl_group_bundle_permission.c.permission
-                        >= GROUP_OBJECT_PERMISSION_READ,  # Match the uuid of the parent
-                    )
-                )
+            access_via_group = and_(
+                or_(  # Join constraint (group)
+                    cl_group_bundle_permission.c.group_uuid
+                    == self.public_group_uuid,  # Public group
+                    cl_group_bundle_permission.c.group_uuid.in_(
+                        alias(
+                            select([cl_user_group.c.group_uuid]).where(
+                                cl_user_group.c.user_id == user_id
+                            )
+                        )
+                    ),  # Private group
+                ),
+                cl_group_bundle_permission.c.permission
+                >= GROUP_OBJECT_PERMISSION_READ,  # Match the uuid of the parent
             )
-            clause = and_(clause, or_(access_via_owner, access_via_group))
 
+            where_clause = and_(where_clause, or_(access_via_owner, access_via_group))
+
+        table = cl_bundle
+        for join_table, join_condition in zip(join_tables, join_conditions):
```

Comment:
Instead of using parallel arrays, which might be error prone, consider making a new `dataclass` with two fields (the table and the join).

Reply:
Done. But see the comment above about INNER JOIN vs LEFT OUTER JOIN. I left this and the other subquery in the .floating case because they were harder to convert and subsequently forgot to convert them. Sorry about that!